lib/badge.rb

```diff
@@ -0,0 +1,100 @@
+class Badge
+  include DSL
+
+  attr_reader :title, :description, :goal, :rank, :user
+
+  def initialize(user)
+    @user = user
+    describe_context
+    @bestowment = check_bestowment(@rank)
+  end
+
+  def check_bestowment(rank = nil)
+    bestowment = Bestowment.where(
+      badge_id: self.class,
+      user: @user
+    )
+    bestowment = bestowment.where(rank: rank) if has_progress?
+    bestowment.first
+  end
+
+  def bestowed(rank = nil)
+    if has_progress?
+      state = self.class.const_get("Rank#{rank}")
+    else
+      state = self
+    end
+    Bestowment.create(
+      badge_id: self.class,
+      rank: rank,
+      user: @user,
+      bestowed_at: DateTime.now,
+      title: state.title,
+      description: state.description
+    )
+  end
+
+  def progress
+    if self.class.progress.nil?
+      nil
+    else
+      @progress ||= instance_eval(&self.class.progress)
+    end
+  end
+
+  def earned?
+    return false unless @goal
+    if progress.nil?
+      instance_eval(&@goal)
+    else
+      progress > @goal
+    end
+  end
+
+  def run
+    # bestowed badge if all ranks was earned
+    bestowed(@rank) if earned? && @bestowment.nil?
+    # bestowed previous rank if it not earned
+    prev = @rank - 1
+    if prev.positive? && has_progress? && check_bestowment(prev).nil?
+      bestowed(@rank - 1)
+    end
+  end
+
+  def has_progress?
+    self.class.ranks.present?
+  end
```

Comment:
`progress` is not dependent on whether a badge has `ranks`.  `progress` is a completely different thing.